test-runner/src/Editor.js

```diff
@@ -110,47 +110,76 @@ const commandHistoryState = StateField.define({
 });
 
 class ErrorGutterMarker extends GutterMarker {
-  constructor(errorData) {
+  constructor(data) {
     super();
-    this.errorData = errorData;
+    this.data = data;
   }
 
   toDOM() {
-    return document.createTextNode("🛑");
+    return document.createTextNode("🔴");
   }
 }
-const underlineMark = Decoration.mark({ class: "cm-underline" });
+class WarnGutterMarker extends GutterMarker {
+  constructor(data) {
+    super();
+    this.data = data;
+  }
+
+  toDOM() {
+    return document.createTextNode("⚠️");
+  }
+}
+const underlineErrorMark = Decoration.mark({ class: "cm-underline-error" });
+const underlineWarningMark = Decoration.mark({ class: "cm-underline-warning" });
 
-const errorEffect = StateEffect.define({});
-const errorState = StateField.define({
+const consoleEffect = StateEffect.define({});
+const consoleState = StateField.define({
   create() {
     return RangeSet.empty;
   },
   update(value, transaction) {
     value = value.map(transaction.changes);
 
-    const errorEffects = transaction.effects.filter((error) =>
-      error.is(errorEffect)
+    const consoleEffects = transaction.effects.filter((error) =>
+      error.is(consoleEffect)
     );
-    value = value.update({
-      add: errorEffects.map((error) => {
-        return new ErrorGutterMarker(error.value).range(
-          error.value.from,
-          error.value.to
-        );
-      }),
-      sort: true,
-    });
+    if (consoleEffects.length > 0) {
```

Comment:
This new check plus the line below (148 `filter: () => false`) means that we:
1. Only update this when the effect update contains a consoleEffect
2. Remove all of the existing consoleEffect, and adds them all back

Prior to this we had to make sure to only add the new consoleEffects or else we would double add them. This filter, then add approach is better for our case since we maintain the total list of consoleEffects outside of the editor and pass it in every time.